src/pip/_internal/models/format_control.py

```diff
@@ -23,7 +27,14 @@ def __init__(self, no_binary=None, only_binary=None):
 
     def __eq__(self, other):
         # type: (object) -> bool
-        return self.__dict__ == other.__dict__
+        if isinstance(other, self.__class__):
+            if self.__slots__ == other.__slots__:
+                attr_getters = [operator.attrgetter(attr)
+                                for attr in self.__slots__]
+                return all(getter(self) == getter(other)
+                           for getter in attr_getters)
```

Comment:
Equality on list could have ordering issues; we probably need to cast the result; something like

```python
self_attrs = {k: getattr(self, k) for k in self.__slots__}
other_attrs = {k: getattr(other, k) for k in other.__slots__}
return self_attrs == other_attrs
```

Reply:
We're protected by a `self.__slots__ == other.__slots__`.

Reply:
> We're protected by a `self.__slots__ == other.__slots__`, which is reasonable with `self.__class__` IMO.

Subclassing may result in `self.__class__ != other.__class__` (therefore not guarenteeing the `__slots__` declarations match), which is the main issue I have with the current implementation.